text/metrics/0238-pipeline-monitoring.md

```diff
@@ -0,0 +1,495 @@
+or removed from the pipeline. Any component that removes an item from the 
+pipeline should increment the outgoing item instrument. An item being removed 
+from the pipeline doesn't indicate an error (an example of this is sampling). 
+
+In the context of an OpenTelemetry Collector pipeline, mostly exporters will
+be incrementing outgoing items as they export them. It's also possible that a
+processor or connector could discard items. Any component could encounter an
+error leading to removing items from the pipeline.
+
+### Outcomes
+
+Below are the possible outcomes for incoming and outgoing items. 
+
+#### Successful Outcomes
+
+- `accepted`: Indicates a normal, synchronous request success case.
+  The item was received by the next segment of the pipeline, which
+  returned success.
+- `discarded`: Indicates a successful outcome in which the next stage
+  of the pipeline does not handle the event, as by a sampling
+  processor.
+
+#### Error Outcomes: Transient
+
+- `dropped`: The component introduced an original failure and did not
+  send to the next stage in the pipeline.
+- `timeout`: The item was in the process of being sent but the request
+  timed out, or its deadline was exceeded.  In this case, it is
+  undetermined whether the follower saw the item or not.
+- `exhausted`: The item was handled by the next stage of the pipeline,
+  which returned an error code indicating that it was overloaded.  If
+  the resource being exhausted is local and the item was not handled
+  by the next stage of the pipeline, record the item `dropped` and
+  return a resource-exhausted status code to the preceding component, who will
+  record an `exhausted` outcome for incoming items.
+- `retryable`: The item was handled by the next stage of the pipeline,
+  which returned a retryable error status not covered by any of the
+  above values.
+
+#### Error Outcomes: Permanent
+
+- `rejected`: The item was handled by the next stage of the pipeline,
+  which returned a permanent error status or partial success status
+  indicating that some items could not be accepted.
+- `unknown`: May be used when the component is suppressing errors and
+  not actually counting successes and failures. As a special case,
+  the outcome `deferred:unknown` indicates that a success response
+  was given and no information about the actual outcome is available.
+
+#### Error Outcomes: Deferred
+
+- `deferred:<failure outcome>`: Deferred cases are where the preceder receives 
+  a success response and the true outcome is failure, but this is not known 
+  until later. The item is counted as `deferred:` combined with the failure 
+  outcome that would otherwise have been counted. The failure outcome could be 
+  transient or permanent. An example of when deferred would be used is in an 
+  asynchronous Collector pipeline.
+
+Additional examples of these outcomes can be found in the Appendix.
+
+### Collector Pipelines With Multiple Exporter Components
```

Comment:
There are some other arrangements of components which I'm trying to fit to the proposed model. Some of them may be worth including in this document as well. Here's how I'm understanding them:
1. _A single collector pipeline with multiple exporters_. As already noted, this includes a fanout point. The document describes how to aggregate outcomes from multiple exporters, effectively ensuring that `Incoming(Segment) == Outgoing(Segment)`.
2. _A single collector pipeline with multiple receivers_. This is fairly straightforward and common. The incoming items are just summed together. Probably doesn't require a dedicated section but we could include it for symmetry.
3. _A single receiver shared by multiple collector pipelines_. This is the other type of fanout point in the collector. From here there are some similar considerations to the first case, but instead of fanning out to multiple exporters, we instead fanout to entire pipelines. What is the relationship between an item arriving at such a receiver and the outcomes of passing it to multiple pipelines? For example, one pipeline may successfully export the item while the other encounters an error which propagates back to the receiver. Does this count as "origin:received" for both pipelines and then each pipeline show a different outcome?
4. _A single exporter shared by multiple collector pipelines_. This is the other type of merge point in the collector. In this case, I think synchronous outcomes can resolve back to a specific pipeline, but async outcomes may not be related to any specific pipeline. For example, if 2 pipelines each send 10 items to an exporter, which then batches all 20 items together into a single export request, the outcome may be "20 deferred:rejected", but it would be incorrect for either pipeline to incorporate that count directly. Is there any way to handle this? Otherwise, maybe this is just a caveat for interpreting deferred outcomes.

Reply:
I wonder about adding to alternatives considered:

[The collector's logic for `fanoutconsumer` uses `multierr`](https://github.com/open-telemetry/opentelemetry-collector/blob/effe26710d251afb58ce4efc5ae5aac52d9b61a0/internal/fanoutconsumer/traces.go#L47), building on Go's `Unwrap() []error` idiom to enclose all the errors. In my opinion, it would be nice to see `fanoutconsumer` manage the logic of deciding how to transform multiple errors into a single error, so that we could specify that fanoutconsumer dictates the N-to-1 problem and the observability mechanism just follows whatever it decides.